### PR TITLE
instrument tracing and add metrics for Azure SDK requests

### LINKED ISSUE #311: Add tracing and metrics - Open Telemetry
/kind feature
Add tracing and metrics with Open Telemetry so CAPZ is observable. 
Logging cluster provisioning metrics which will eventually export to Azure monitor. 

cc. @devigned

### PR DESCRIPTION
**Pull request by devigned**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:
This PR enables tracing through OpenTelemetry and Azure SDK metrics.
- `make tilt-up` now creates a kind cluster with ingress enabled which Jaeger needs to expose the UI without a tunnel. By default Jaeger is deployed as an operator and the capz controller is annotated to have the Jaeger all-in-one sidecar added (`localhost:80`).
- `azurecluster-controller` and everything below it is instrumented for tracing.

tracing

Todos:
- Add Prometheous to tilt so it's super easy to see metrics

/hold

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #311

**TODOs**:

- [ ] squashed commits
- [x] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
enable Azure SDK metrics and tracing
```

### CODE REVIEW

**Comment by CecileRobertMichon** [comment]
🎉 

What do you think about adding a metrics section to the development doc to explain how to add and use the metrics for development?

**Comment by devigned** [comment]
/hold cacel

I think this is ready for a look when folks have a minute. I know the change set is really big... I'm sorry. I've tried to break it up with docs, tracing, metrics and test `context.Context` change batches.

The majority of the additional lines is from the jaeger-operator and prometheus operator. I didn't think the PR would be all that useful without being able to have a working telemetry environment in KinD, so I included the operators in the hack directory.

**Review by nader-ziada — COMMENTED** [review]
went through all the changes to make sure the names and functions match and all is good, will still try to run it 👍

**Review by CecileRobertMichon — APPROVED** [review]
looks great. Tried it out using Tilt on this branch and it's working as expected. Also went through all the files looking for possible typos in spans but didn't find any 💯

File: config/kustomization.yaml [hunk 1]
```diff
@@ -3,7 +3,7 @@ namePrefix: capz-
 # Labels to add to all resources and selectors.
 commonLabels:
   cluster.x-k8s.io/provider: "infrastructure-azure"
-bases:
+resources:
```
[the remark this answers is not in the document]
Comment (devigned) [reply config/kustomization.yaml:6]: bases is deprecated. https://github.com/kubernetes-sigs/kustomize/blob/2a8a17e3af263c8cd61452bd2fc5e2e7d7bc5330/api/types/kustomization.go#L85-L88

**Comment by devigned** [comment]
Looks like the IPv6 e2e test is failing to get the 3rd control plane node to become ready... I'm looking into it, but I don't think this PR would have any ill effects on IPv6 clusters.

**Pull request merged**